src/resolution_functions/instrument.py

```diff
@@ -809,16 +924,58 @@ def default_option_for_configuration(self, model_name: str, configuration: str)
             If the provided `configuration` is not supported for the `model_name` model of this
             instrument.
         """
-        try:
-            model = self._models[model_name]
-        except KeyError:
-            raise InvalidModelError(model_name, self)
-
-        configurations = model['configurations']
+        configurations = self._resolve_model(model_name)['configurations']
 
         try:
             configurations = configurations[configuration]
         except KeyError:
             raise InvalidConfigurationError(configuration, model_name, self)
 
         return configurations['default_option']
+
+
+def _format_table(contents: list[list[str | None]],
+                  longest: list[int],
+                  padding: int = 4) -> str:
+    """
+    Formats `contents` into a table.
+
+    Parameters
+    ----------
+    contents
+        The data to be formatted into a table. Each entry in the list of lists will be turned into
+        a single cell. Strings are placed into cells (with `padding` applied) while `None` values
+        are turned into empty cells.
+    longest
+        The length of the longest entry in each column.
+    padding
+        Extra padding to apply to the entries. Default is 4.
+
+    Returns
+    -------
+    table_str
+        A string containing the table.
+    """
+    longest = np.array(longest) + padding
+    out = []
+    separator = '|' + '|'.join(['-' * (i + 1) for i in longest]) + '|'
+
+    for i, line in enumerate(contents):
+        new_line = '| '
+        for val, length in zip(line, longest):
+            if val is None:
+                new_line += ' ' * length + '| '
+            else:
+                new_line += val + ' ' * (length - len(val)) + '| '
+
+        if line[0] is not None:
+            if i == 1:
+                out.append(separator.replace('-', '='))
+            else:
+                out.append(separator)
+
+        out.append(new_line)
+
+    out.append(separator)
+
+    return '\n'.join(out)
```

Comment:
Maybe it's just my functional-programming brain, but you can avoid a lot of this fuss by using Python's format functions. It needs some hackery to get the *same* table, but a similar table could be done without incrementing `padding` and `preproc`, different formats, etc. `sep_nl` is just because you can't have `\` in an f-string. 
```suggestion
    padding = padding + 1
    preproc = [["-" if elem is None else elem for elem in row] for row in contents]

    sep_line = f"|{'|'.join('-' * (i+padding) for i in longest)}|"
    header_sep_line = sep_line.replace("-", "=")
    sep_nl = "\n" + sep_line

    data_line = "|" + "|".join(f"{{:^{length+padding}}}" for length in longest) + "|"
    formatted = map(lambda x: data_line.format(*x), preproc)

    out = f"""\
{sep_line}
{next(formatted)}
{header_sep_line}
{sep_nl.join(formatted)}
{sep_line}
    """

    return out
```

Reply:
That's just because I stuck in centred (`^`) rather than left aligned (`<`) in the formats

Reply:
Change `data_line` to:
```python
    data_line = "| " + "| ".join(f"{{:<{length+padding}}}" for length in longest) + "|"
```
And it should do what you had before.

Reply:
Ah! I see what you mean, it was grouping things previously.

Reply:
Makes `more_itertools` a required dep, but:
```suggestion
    padding = padding + 1

    preproc = [["" if elem is None else elem for elem in row] for row in contents]

    sep_line = f"|{'|'.join('-' * (i+padding) for i in longest)}|"
    header_sep_line = sep_line.replace("-", "=")
    sep_nl = "\n" + sep_line + "\n"
    data_line = "| " + "| ".join(f"{{:<{length+padding-1}}}" for length in longest) + "|"

    it = more_itertools.split_before(preproc, lambda x: x[0])
    formatted = ('\n'.join(map(lambda x: data_line.format(*x), blk)) for blk in it)

    out = f"""\
{sep_line}
{next(formatted)}
{header_sep_line}
{sep_nl.join(formatted)}
{sep_line}
    """

    return out
```

Reply:
That does indeed work! @ajjackson what do you think?

Reply:
> I like the approach and it's easier to quickly see what is going on. Some more verbose variable names might help further. (e.g. `preproc -> contents_as_strings`) Maybe a comment above #966
> 
> ```
> # e.g. "|{:<10}|{:<10}|{:<12}|{:<16}|" for str.format()
> ```
> 
> Instead of adding the dependency we could implement a function here like:
> 
> ```python
> def split_on_model(contents: Iterator[list[str]]) -> Iterator[list[list[str]]]:
>     """Break the table lines into chunks where the first column is occupied"""
>     group = [next(contents)]
>     for line in contents:
>         if line[0]:
>             yield group
>             group = []
>         else:
>             group.append(line)
>     if group:
>         yield group
> ```
> 
> (I haven't tested this, typed straight into github website...)

You'll need `group = [line]` to avoid dropping the head.  That could also handle the preprocessing of blanks.

Reply:
So, do you want to go ahead with this? With both `split_on_model` and `_format_table` in a new utils module?

Reply:
Are they likely to be re-used? If so then collecting some generic output-printing utils makes sense. This kind of modularity is good for testing.

But if they would have private names and be used in one place, I'm not sure it's so beneficial to split them into a new module.

Reply:
Not at the moment, no. In the future, difficult to say, but I can't think of anything

Reply:
Could you give me an example of how the `split_on_model` is supposed to be used inside `_format_table`? Can't seem to figure out how it's supposed to replace `more_itertools.split_before`